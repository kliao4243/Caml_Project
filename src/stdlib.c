#include <stdio.h>
#include <stdlib.h> 
#include <time.h>

struct _track{
	int instrument;
	char** melody;
	int* rhythm;
};

int pitch_to_int(char* s){
    char a = s[0];
    char b = s[1];
    char c = s[2];
    int result = 23;
    if(a-'0' < 4){
        result += 2 * (a-'0') - 1;
    }else{
        result += 2 * (a-'0') - 2;
    }
    switch(b){
        case 'b':
            result--;
            break;
        case '#':
            result++;
            break;
    }
    result += (c-'1') * 12;
    return result;
}

int ran(int range){
    return (rand() % range);
}
