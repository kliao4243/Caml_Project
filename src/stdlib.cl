struct _track{
	int instrument;
	int size;
	Pitch[20] melody;
	int[20] rhythm;
};
_track build_track(int instrument, int size, Pitch[] melody, int[] rhythm){
	_track res;
	int i;
	Pitch[20] new_melody = [1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1,1b1];
	int[20] new_rhythm = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0];
	res.instrument = instrument;
	res.size = size;
	for(i = 0; i < size; i = i + 1){
		new_melody[i] = melody[i];
		new_rhythm[i] = rhythm[i];
	} 
	res.melody = new_melody;
	res.rhythm = new_rhythm;
	return res;
}

void generate_music(_track t){
	int i;
	print("track_start"); /*size of tracks*/
	print(t.instrument);
	print(t.size);
	for(i = 0; i < t.size; i = i + 1){
		print(pitch_to_int(t.melody[i]));
		print(t.rhythm[i]);
	}
	return;
}
