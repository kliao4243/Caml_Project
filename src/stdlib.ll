; ModuleID = 'stdlib.c'
source_filename = "stdlib.c"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx10.14.0"

; Function Attrs: noinline nounwind optnone ssp uwtable
define i8* @up_8(i8*) #0 {
  %2 = alloca i8*, align 8
  store i8* %0, i8** %2, align 8
  %3 = load i8*, i8** %2, align 8
  ret i8* %3
}

; Function Attrs: noinline nounwind optnone ssp uwtable
define i8* @down_8(i8*) #0 {
  %2 = alloca i8*, align 8
  store i8* %0, i8** %2, align 8
  %3 = load i8*, i8** %2, align 8
  ret i8* %3
}

; Function Attrs: noinline nounwind optnone ssp uwtable
define i32 @pitch_to_int(i8*) #0 {
  %2 = alloca i32, align 4
  %3 = alloca i8*, align 8
  %4 = alloca i8, align 1
  %5 = alloca i8, align 1
  %6 = alloca i8, align 1
  %7 = alloca i32, align 4
  store i8* %0, i8** %3, align 8
  %8 = load i8*, i8** %3, align 8
  %9 = getelementptr inbounds i8, i8* %8, i64 0
  %10 = load i8, i8* %9, align 1
  %11 = sext i8 %10 to i32
  %12 = icmp eq i32 %11, 49
  br i1 %12, label %13, label %26

; <label>:13:                                     ; preds = %1
  %14 = load i8*, i8** %3, align 8
  %15 = getelementptr inbounds i8, i8* %14, i64 1
  %16 = load i8, i8* %15, align 1
  %17 = sext i8 %16 to i32
  %18 = icmp eq i32 %17, 98
  br i1 %18, label %19, label %26

; <label>:19:                                     ; preds = %13
  %20 = load i8*, i8** %3, align 8
  %21 = getelementptr inbounds i8, i8* %20, i64 2
  %22 = load i8, i8* %21, align 1
  %23 = sext i8 %22 to i32
  %24 = icmp eq i32 %23, 49
  br i1 %24, label %25, label %26

; <label>:25:                                     ; preds = %19
  store i32 0, i32* %2, align 4
  br label %73

; <label>:26:                                     ; preds = %19, %13, %1
  %27 = load i8*, i8** %3, align 8
  %28 = getelementptr inbounds i8, i8* %27, i64 0
  %29 = load i8, i8* %28, align 1
  store i8 %29, i8* %4, align 1
  %30 = load i8*, i8** %3, align 8
  %31 = getelementptr inbounds i8, i8* %30, i64 1
  %32 = load i8, i8* %31, align 1
  store i8 %32, i8* %5, align 1
  %33 = load i8*, i8** %3, align 8
  %34 = getelementptr inbounds i8, i8* %33, i64 2
  %35 = load i8, i8* %34, align 1
  store i8 %35, i8* %6, align 1
  store i32 23, i32* %7, align 4
  %36 = load i8, i8* %4, align 1
  %37 = sext i8 %36 to i32
  %38 = sub nsw i32 %37, 48
  %39 = icmp slt i32 %38, 4
  br i1 %39, label %40, label %48

; <label>:40:                                     ; preds = %26
  %41 = load i8, i8* %4, align 1
  %42 = sext i8 %41 to i32
  %43 = sub nsw i32 %42, 48
  %44 = mul nsw i32 2, %43
  %45 = sub nsw i32 %44, 1
  %46 = load i32, i32* %7, align 4
  %47 = add nsw i32 %46, %45
  store i32 %47, i32* %7, align 4
  br label %56

; <label>:48:                                     ; preds = %26
  %49 = load i8, i8* %4, align 1
  %50 = sext i8 %49 to i32
  %51 = sub nsw i32 %50, 48
  %52 = mul nsw i32 2, %51
  %53 = sub nsw i32 %52, 2
  %54 = load i32, i32* %7, align 4
  %55 = add nsw i32 %54, %53
  store i32 %55, i32* %7, align 4
  br label %56

; <label>:56:                                     ; preds = %48, %40
  %57 = load i8, i8* %5, align 1
  %58 = sext i8 %57 to i32
  switch i32 %58, label %65 [
    i32 98, label %59
    i32 35, label %62
  ]

; <label>:59:                                     ; preds = %56
  %60 = load i32, i32* %7, align 4
  %61 = add nsw i32 %60, -1
  store i32 %61, i32* %7, align 4
  br label %65

; <label>:62:                                     ; preds = %56
  %63 = load i32, i32* %7, align 4
  %64 = add nsw i32 %63, 1
  store i32 %64, i32* %7, align 4
  br label %65

; <label>:65:                                     ; preds = %56, %62, %59
  %66 = load i8, i8* %6, align 1
  %67 = sext i8 %66 to i32
  %68 = sub nsw i32 %67, 49
  %69 = mul nsw i32 %68, 12
  %70 = load i32, i32* %7, align 4
  %71 = add nsw i32 %70, %69
  store i32 %71, i32* %7, align 4
  %72 = load i32, i32* %7, align 4
  store i32 %72, i32* %2, align 4
  br label %73

; <label>:73:                                     ; preds = %65, %25
  %74 = load i32, i32* %2, align 4
  ret i32 %74
}

; Function Attrs: noinline nounwind optnone ssp uwtable
define i32 @ran(i32) #0 {
  %2 = alloca i32, align 4
  store i32 %0, i32* %2, align 4
  %3 = call i32 @rand()
  %4 = load i32, i32* %2, align 4
  %5 = srem i32 %3, %4
  ret i32 %5
}

declare i32 @rand() #1

attributes #0 = { noinline nounwind optnone ssp uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }

!llvm.module.flags = !{!0, !1, !2}
!llvm.ident = !{!3}

!0 = !{i32 2, !"SDK Version", [2 x i32] [i32 10, i32 14]}
!1 = !{i32 1, !"wchar_size", i32 4}
!2 = !{i32 7, !"PIC Level", i32 2}
!3 = !{!"Apple LLVM version 10.0.1 (clang-1001.0.46.3)"}
