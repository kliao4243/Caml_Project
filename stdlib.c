#include <stdio.h>
struct _track{
	int instrument;
	char** melody;
	int* rhythm;
};

int print_arr(int a){
    int res = a + 100;
    printf("%d\n",res);
	return 0;
}
int pitch_to_int(char* s){
    char a = s[0];
    char b = s[1];
    char c = s[2];
    int result = 23;
    if(a-'0' < 4){
        result += 2 * (a-'0') - 1;
    }else{
        result += 2 * (a-'0') - 2;
    }
    switch(b){
        case 'b':
            result--;
            break;
        case '#':
            result++;
            break;
    }
    result += (c-'1') * 12;
    return result;
}

