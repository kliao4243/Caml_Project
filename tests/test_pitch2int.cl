/* author: Yuanji Huang */
#include "src/stdlib.cl";
int main(){
    int a_1;
    int b_1;
    int c_1;
    Pitch a = 4^4;
    Pitch b = 1#7;
    Pitch c = 2b3;
    a_1 = pitch_to_int(a);
    b_1 = pitch_to_int(b);
    c_1 = pitch_to_int(c);
    print(a_1);
    print(b_1);
    print(c_1);
}
