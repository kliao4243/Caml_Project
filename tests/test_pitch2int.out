65
97
49
